benchmarks/benchmarks.cpp

```diff
@@ -0,0 +1,448 @@
+//////////////////////////////////////////////////////////////////////////////////////
+//
+// (C) Daniel Strano 2017, 2018. All rights reserved.
+//
+// This is a multithreaded, universal quantum register simulation, allowing
+// (nonphysical) register cloning and direct measurement of probability and
+// phase, to leverage what advantages classical emulation of qubits can have.
+//
+// Licensed under the GNU General Public License V3.
+// See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
+// for details.
+
+#include <atomic>
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "catch.hpp"
+#include "qfactory.hpp"
+
+#include "benchmarks.hpp"
+
+using namespace Qrack;
+
+#define EPSILON 0.001
+#define REQUIRE_FLOAT(A, B) do {                                            \
+        double __tmp_a = A;                                                 \
+        double __tmp_b = B;                                                 \
+        REQUIRE(__tmp_a < (__tmp_b + EPSILON));                             \
+        REQUIRE(__tmp_b > (__tmp_b - EPSILON));                             \
+    } while (0);
+
+void benchmarkLoop(std::shared_ptr<std::default_random_engine> rng, QInterfaceEngine engineType, QInterfaceEngine subEngineType,std::function<void(QInterfacePtr, int)> fn) {
+
+    const int ITERATIONS = 100;
+
+    std::cout<<std::endl;
+    std::cout<<ITERATIONS<<" iterations";
+    std::cout<<std::endl;
+    std::cout<<"# of Qubits, ";
+    std::cout<<"Average Time (ms), ";
+    std::cout<<"Std. Error (ms)"<<std::endl;
+
+    clock_t start, t, all;
+    clock_t trials[ITERATIONS];
+
+    int i, j;
+
+    double avgt, stdet;
+
+    // Grover's search inverts the function of a black box subroutine.
+    // Our subroutine returns true only for an input of 100.
+    for (j = 3; j <= 20; j++) {
+        QInterfacePtr qftReg = CreateQuantumInterface(engineType, subEngineType, j, 0, rng);
+        start = clock();
+        all = start;
+        t = start;
+        avgt = 0.0;
+        for (i = 0; i < ITERATIONS; i++) {
+
+            // Run loop body
+            fn(qftReg, j);
+
+            // Display timing
+            t = clock() - all;
+            trials[i] = t;
+            avgt += t;
+            all = clock();
+        }
+        all = clock() - start;
+        avgt /= ITERATIONS;
+
+        stdet = 0.0;
+        for (i = 0; i < ITERATIONS; i++) {
+            stdet += (trials[i] - avgt) * (trials[i] - avgt);
+        }
+        stdet = sqrt(stdet / ITERATIONS);
+
+        std::cout<<(int)j<<", ";
+        std::cout<<(avgt * 1000.0 / CLOCKS_PER_SEC)<<", ";
+        std::cout<<(stdet * 1000.0 / CLOCKS_PER_SEC)<<std::endl;
+    }
+}
+
+TEST_CASE_METHOD(QInterfaceTestFixture, "test_cnot")
```

Comment:
Don’t use TEST_CASE_METHOD - that creates (and destroys) a qftReg for each call.  Just use TEST_CASE (I think - look at the par_for tests for an example)

Reply:
Actually, I took out the default qftReg in the benchmark variant, so it doesn't incur that overhead. The only unit is created in the loop function.